accepted/component-model-async.md

```diff
@@ -0,0 +1,587 @@
+passed to another component but remain available in the original component until
+explicitly dropped.
+      
+## Host APIs for creating, using, and sharing `stream`s, `future`s, and `error`s
+
+Ideally, the public API for `stream`s and `future`s would look a lot like the
+[futures::channel::mpsc](https://docs.rs/futures/latest/futures/channel/mpsc/index.html)
+and
+[futures::channel::oneshot](https://docs.rs/futures/latest/futures/channel/oneshot/index.html)
+APIs, respectively.  Or perhaps
+[AsyncRead](https://docs.rs/futures/latest/futures/io/trait.AsyncRead.html) and
+[AsyncWrite](https://docs.rs/futures/latest/futures/io/trait.AsyncWrite.html)
+APIs for the special case of `stream<u8>`.  However, we have the same challenge
+here as with `call_concurrent`: each of these types need a `StoreContextMut<T>`
+to do their job, i.e. lifting from and lowering to Wasm linear memory (which may
+involve `cabi_realloc` calls, etc.).  Therefore, as with `call_concurrent`, we
+use the `Promise` type described above to achieve `Store`-friendly concurrency.
+
+```rust
+pub struct FutureSender<T> { ... }
+
+impl<T> FutureSender<T> {
+    pub fn send(self, value: T) -> Promise<Result<()>>
+    where
+        T: func::Lower + Send + Sync + 'static,
+    { ... }
+
+    pub fn close<U, S: AsContextMut<Data = U>>(self, store: S) -> Result<()> { ... }
+}
+
+pub struct FutureReceiver<T> { ... }
+
+impl<T> FutureReceiver<T> {
+    pub fn receive(self) -> Promise<Result<Option<T>>>
+    where
+        T: func::Lift + Sync + Send + 'static,
+    { ... }
+
+    pub fn close<U, S: AsContextMut<Data = U>>(self, store: S) -> Result<()> { ... }
+}
+
+pub fn future<T, U, S: AsContextMut<Data = U>>(
+    store: S,
+) -> Result<(FutureSender<T>, FutureReceiver<T>)> { ... }
+
+pub struct StreamSender<T> { ... }
+
+impl<T> StreamSender<T> {
+    pub fn send(&mut self, values: Vec<T>) -> Promise<Result<()>>
+    where
+        T: func::Lower + Send + Sync + 'static,
+    { ... }
+
+    pub fn close<U, S: AsContextMut<Data = U>>(self, store: S) -> Result<()> { ... }
+}
+
+pub struct StreamReceiver<T> { ... }
+
+impl<T> StreamReceiver<T> {
+    pub fn receive(&mut self) -> Promise<Result<Option<Vec<T>>>
+    where
+        T: func::Lift + Sync + Send + 'static,
+    { ... }
+
+    pub fn close<U, S: AsContextMut<Data = U>>(self, store: S) -> Result<()> { ... }
+}
+
+pub fn stream<T, U, S: AsContextMut<Data = U>>(
+    store: S,
+) -> Result<(StreamSender<T>, StreamReceiver<T>)> { ... }
+```
+
+The `error` API, on the other hand, is quite simple since you can't do anything
+with it except pass it around:
+
+```rust
+pub struct Error { ... }
+
+pub fn error<U, S: AsContextMut<Data = U>>(store: S) -> Result<Error> { ... }
+```
+
+We expect that `error`s will eventually have `anyhow`-style functionality
+(diagnostic messages, backtraces, downcast-able payloads, etc.), but for now
+they're just opaque handles.
+
+# Rationale and alternatives
```

Comment:
I think to some degree we don't really need this kind of discussion: the "what else could we do" question has been answered at the spec level, and we'd have had a ton of opportunity to voice disagreement there. At this point, I think we should handle this the same as other features, such as the recent support for [Tail Calls](https://github.com/bytecodealliance/rfcs/blob/main/accepted/tail-calls.md#rationale-and-alternatives).

As such, I wonder if there are possible alternative approaches to implementing support of the spec feature that would be worth discussing here? And if not, is there a way to explain why that is the case as part of a "there really isn't much of an alternative" paragraph?